teuthology/provision/maas.py

```diff
@@ -0,0 +1,430 @@
+
+
+def get_session():
+    """Create an OAuth1Session for communicating with the MaaS server"""
+    if not enabled():
+        raise RuntimeError("MAAS is not configured!")
+
+    key, token, secret = config.maas["api_key"].split(":")
+    return OAuth1Session(
+        key,
+        resource_owner_key=token,
+        resource_owner_secret=secret,
+        signature_method=SIGNATURE_PLAINTEXT
+    )
+
+
+class MAAS(object):
+    """Reimage machines with https://maas.io"""
+
+    def __init__(self, name,  os_type=None, os_version=None):
+        """Initialize the MAAS object
+
+        :param name: The fully-qualified domain name of the machine to manage
+        :param os_type: The OS type to deploy (e.g. "ubuntu")
+        :param os_version: The OS version to deploy (e.g. "22.04")
+        """
+        self.session = get_session()
+
+        self.remote = teuthology.orchestra.remote.Remote(
+            misc.canonicalize_hostname(name))
+        self.name = self.remote.hostname
+        self.shortname = self.remote.shortname
+
+        self.log = log.getChild(self.shortname)
+
+        self.os_type, self.os_version, self.system_id = self._get_system_info(
+            os_type, os_version)
+
+    def _get_system_info(self, os_type, os_version):
+        """Get the system info string for the deployed OS
+
+        :param os_type: The OS type as a string
+        :param os_version: The OS version as a string
+        :returns: os type, version, and system id as strings
+        """
+        _os_type, _os_version = None, None
+
+        maas_machine = self.get_machines_data()
+        if maas_machine.get("status_name").lower() == "ready":
+            if not (os_type and os_version):
+                raise RuntimeError(
+                    "Machine %s is not deployed, please provide "
+                    "OS type and version" % self.shortname)
+            _os_type, _os_version = os_type.lower(), str(os_version)
+        elif maas_machine.get("status_name").lower() == "deployed":
+            if os_type and os_version:
+                _os_type, _os_version = os_type.lower(), str(os_version)
+            else:
+                _os_type = maas_machine.get("osystem", "").lower()
+                _os_version = self._get_os_version(
+                    os_type, maas_machine.get("distro_series", "").lower()
+                )
+        else:
+            raise RuntimeError(
+                "MaaS machine %s is not ready or deployed, current status: %s"
+                % (self.shortname, maas_machine.get("status_name"))
+            )
+
+        return _os_type, _os_version, maas_machine.get("system_id")
+
+    def _get_os_version(self, os_type, os_codename):
+        """Get the OS version of the deployed OS
+
+        :param os_type: The OS type as a string
+        :param os_codename: The OS codename as a string
+        :returns: The OS version as a string
+        """
+        if os_type != "ubuntu":
+            return os_codename
```

Comment:
This is confusing, the `_get_os_version()` should return a version, here it is returning _codename_. And what is gonna return for none ubuntu here?